### ISSUE TITLE
Create and implement OnlineSigner interface

### ISSUE THREAD

**Issue by webmaster128** [CosmWasm/cosmjs#316]
See https://github.com/CosmWasm/cosmjs/issues/300#issuecomment-661910637 and the following comments

**Comment by Thunnini** [comment]
It seems that @ethanfrey's `online signer` suggestion is general enough and in a glace, there is no problem in implementing the `online signer` for Keplr. 
``` typescript
export interface OnlineSigner {
  /*
   * must call once and succeed before using the signer (login / authorization step)
   */
  readonly enable: () => Promise<boolean>;

  /*
   * Same as with OfflineSigner
   */
  readonly getAccounts: () => Promise<readonly AccountData[]>;

  /**
   * Signs with whichever key corresponds to provided bech32-encoded address. Rejects if not enabled.
   * Will query chain for account_number and sequence number if not set by caller.
   * Will auto-set fee if not set by caller  
   * Will submit to the blockchain and return the TxHash if it passes `CheckTx` (else error)
   */
  readonly signAndSubmit: (address: string, request: SignRequest) => Promise<TxHash>;
}
```
It permits the wallet to implement its own way in more general. It seems to have a great advantage in that it leaves the wallet to support methods like `getTxBuilderConfig` or not.

Regarding what has been discussed, I've been rethinking about if something like getTxBuilderConfig is necessary.
Initially, I thought that a separate getTxBuilderConfig was necessary as cosmos-sdk deals with tx as an interface and permits the other chain to use other tx struct which is not stdTx. 
But looking at cosmos-sdk blockchains, it seems like there are no other blockchains that doesn't use stdTx.
Also, reexamining this thought, getTxBuildeConfig may not feel intuitive and still has issues with checking if the tx’s actual config matches at the time of tx builder config and signing.
Because those two methods are separate, Keplr requests two pages consisting of a fee-setting page and a signing request. But perhaps this is unnecessary after all.
Since the only stdTx is used, Keplr is able to set the tx config at the time of signing, and that may be more convenient for users.
I’m still thinking through this problem, but the interface above can be implemented regardless of the design decisions I’ve discussed above and the wallet interface part can be tested during wallet implementation. Therefore, I think developing it in this way can help speed up the development process.

**Comment by ethanfrey** [comment]
Thank you for the feedback @Thunnini 

>  But looking at cosmos-sdk blockchains, it seems like there are no other blockchains that doesn't use stdTx.

When I implemented fee delegation (last September), I added a field to the FeeInfo struct, modifying the StdTx with an additional optional field. A number of core devs were uncomfortable with that for some reason and asked me to make my own FeeTx implementation. 2-3 days later I had this working for all go tests, but then discovered there is *absolutely no support* in the cli or rest interface for any Tx but StdTx.

In short, Tx is a nice idea for an interface, but it is near impossible to use anything besides `StdTx`. I think the only people who change this, just fork the sdk and change StdTx. And unlikley to need to handle those cases.

**Comment by ethanfrey** [comment]
If we provided such an interface, would you be happy with a cosmjs provider that fulfilled such interface and allowed the keplr extension to handle fee setting and submission?

**Comment by Thunnini** [comment]
Looks good to me. Once this has been implemented, I will start working on provider injection in parallel. I’ll post an issue if I run into any problems

**Comment by webmaster128** [comment]
The `InProcessOnlineSigner` from #403 shows a cool way to wrap an `OfflineSigner` in an `OnlineSigner`. This should be used when `OnlineSigner` is implemented.
